File: go/internal/fetcher/fetcher.go
package fetcher

import (
	"context"
	"fmt"
	"io"
	"math"
	"math/rand"
	"net/http"
	"sync"
	"time"

	"feedpulse/internal/config"
	"feedpulse/internal/parser"
	"feedpulse/internal/storage"
)

// FetchResult represents the result of fetching a single feed
type FetchResult struct {
	Source       string
	Success      bool
	ItemsCount   int
	NewItems     int
	Error        string
	DurationMs   int64
	Items        []storage.FeedItem
}

// Fetcher handles concurrent feed fetching
type Fetcher struct {
	config *config.Config
	parser *parser.Parser
	client *http.Client
}

// NewFetcher creates a new fetcher instance
func NewFetcher(cfg *config.Config) *Fetcher {
	return &Fetcher{
		config: cfg,
		parser: parser.NewParser(),
		client: &http.Client{
			Timeout: time.Duration(cfg.Settings.DefaultTimeoutSecs) * time.Second,
		},
	}
}

// FetchAll fetches all configured feeds concurrently
func (f *Fetcher) FetchAll(ctx context.Context) []FetchResult {
	// Create a semaphore to limit concurrency
	sem := make(chan struct{}, f.config.Settings.MaxConcurrency)
	
	var wg sync.WaitGroup
	results := make([]FetchResult, len(f.config.Feeds))

	for i, feed := range f.config.Feeds {
		wg.Add(1)
		go func(index int, feed config.Feed) {
			defer wg.Done()

			// Acquire semaphore
			select {
			case sem <- struct{}{}:
				defer func() { <-sem }()
			case <-ctx.Done():
				results[index] = FetchResult{
					Source:  feed.Name,
					Success: false,
					Error:   "cancelled",
				}
				return
			}

			// Fetch the feed
			results[index] = f.fetchFeed(ctx, feed)
		}(i, feed)
	}

	wg.Wait()
	return results
}

// fetchFeed fetches a single feed with retries
func (f *Fetcher) fetchFeed(ctx context.Context, feed config.Feed) FetchResult {
	start := time.Now()

	var lastErr error
	for attempt := 0; attempt <= f.config.Settings.RetryMax; attempt++ {
		if attempt > 0 {
			// Calculate exponential backoff with jitter
			delay := f.calculateBackoff(attempt)
			select {
			case <-time.After(delay):
			case <-ctx.Done():
				return FetchResult{
					Source:     feed.Name,
					Success:    false,
					Error:      "cancelled during retry",
					DurationMs: time.Since(start).Milliseconds(),
				}
			}
		}

		// Attempt to fetch
		data, err := f.fetchURL(ctx, feed)
		if err != nil {
			lastErr = err
			// Don't retry on 404 or client errors
			if isClientError(err) {
				break
			}
			continue
		}

		// Parse the feed
		parseResult := f.parser.Parse(feed.Name, feed.FeedType, data)
		
		// Log parse errors but don't fail
		if len(parseResult.Errors) > 0 {
			for _, parseErr := range parseResult.Errors {
				fmt.Fprintf(io.Discard, "warning: %s: %s\n", feed.Name, parseErr)
			}
		}

		duration := time.Since(start).Milliseconds()
		return FetchResult{
			Source:     feed.Name,
			Success:    true,
			ItemsCount: len(parseResult.Items),
			Items:      parseResult.Items,
			DurationMs: duration,
		}
	}

	// All retries failed
	duration := time.Since(start).Milliseconds()
	errorMsg := "unknown error"
	if lastErr != nil {
		errorMsg = lastErr.Error()
	}

	return FetchResult{
		Source:     feed.Name,
		Success:    false,
		Error:      fmt.Sprintf("failed after %d retries: %s", f.config.Settings.RetryMax, errorMsg),
		DurationMs: duration,
	}
}

// fetchURL performs the actual HTTP request
func (f *Fetcher) fetchURL(ctx context.Context, feed config.Feed) ([]byte, error) {
	req, err := http.NewRequestWithContext(ctx, "GET", feed.URL, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	// Add custom headers
	for key, value := range feed.Headers {
		req.Header.Set(key, value)
	}

	// Set default User-Agent if not provided
	if req.Header.Get("User-Agent") == "" {
		req.Header.Set("User-Agent", "feedpulse/1.0")
	}

	resp, err := f.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("HTTP request failed: %w", err)
	}
	defer resp.Body.Close()

	// Check status code
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, &HTTPError{
			StatusCode: resp.StatusCode,
			Status:     resp.Status,
		}
	}

	// Read response body
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	return data, nil
}

// calculateBackoff calculates exponential backoff with jitter
func (f *Fetcher) calculateBackoff(attempt int) time.Duration {
	baseDelay := float64(f.config.Settings.RetryBaseDelayMs)
	delay := baseDelay * math.Pow(2, float64(attempt-1))
	
	// Add jitter (±25%)
	jitter := delay * 0.25 * (2*rand.Float64() - 1)
	totalDelay := delay + jitter
	
	return time.Duration(totalDelay) * time.Millisecond
}

// HTTPError represents an HTTP error response
type HTTPError struct {
	StatusCode int
	Status     string
}

func (e *HTTPError) Error() string {
	return fmt.Sprintf("HTTP %d: %s", e.StatusCode, e.Status)
}

// isClientError checks if an error is a 4xx client error (don't retry these)
func isClientError(err error) bool {
	if httpErr, ok := err.(*HTTPError); ok {
		return httpErr.StatusCode >= 400 && httpErr.StatusCode < 500
	}
	return false
}

File: rust/src/fetcher.rs
use crate::config::{Config, Feed};
use crate::models::FeedItem;
use crate::parser::Parser;
use reqwest::Client;
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::Semaphore;
use tokio::time::sleep;

#[derive(Debug, Clone)]
pub struct FetchResult {
    pub source: String,
    pub items: Vec<FeedItem>,
    pub new_items: usize,
    pub duration_ms: u64,
    pub error: Option<String>,
}

pub struct Fetcher {
    config: Config,
    client: Client,
}

impl Fetcher {
    pub fn new(config: Config) -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(config.settings.default_timeout_secs))
            .build()
            .expect("Failed to build HTTP client");

        Self { config, client }
    }

    pub async fn fetch_all(&self) -> Vec<FetchResult> {
        let feeds = self.config.feeds.clone();
        let max_concurrency = self.config.settings.max_concurrency;

        println!(
            "Fetching {} feeds (max concurrency: {})...",
            feeds.len(),
            max_concurrency
        );

        let semaphore = Arc::new(Semaphore::new(max_concurrency));
        let mut tasks = Vec::new();

        for feed in feeds {
            let sem = semaphore.clone();
            let client = self.client.clone();
            let retry_max = self.config.settings.retry_max;
            let retry_base_delay = self.config.settings.retry_base_delay_ms;

            let task = tokio::spawn(async move {
                let _permit = sem.acquire().await.unwrap();
                Self::fetch_feed(client, feed, retry_max, retry_base_delay).await
            });

            tasks.push(task);
        }

        let mut results = Vec::new();
        for task in tasks {
            if let Ok(result) = task.await {
                results.push(result);
            }
        }

        results
    }

    async fn fetch_feed(
        client: Client,
        feed: Feed,
        retry_max: usize,
        retry_base_delay: u64,
    ) -> FetchResult {
        let start = Instant::now();
        let source = feed.name.clone();

        for attempt in 0..=retry_max {
            match Self::try_fetch(&client, &feed).await {
                Ok(items) => {
                    let duration_ms = start.elapsed().as_millis() as u64;
                    return FetchResult {
                        source,
                        items,
                        new_items: 0, // Will be updated by storage
                        duration_ms,
                        error: None,
                    };
                }
                Err(e) => {
                    if attempt < retry_max {
                        // Check if we should retry based on error type
                        let should_retry = match &e {
                            FetchError::Http(status) => {
                                // Don't retry 4xx errors except 429
                                if status.as_u16() == 429 {
                                    true
                                } else if status.is_client_error() {
                                    false
                                } else {
                                    true // Retry 5xx and other errors
                                }
                            }
                            _ => true, // Retry network errors, timeouts, etc.
                        };

                        if should_retry {
                            let delay = retry_base_delay * 2_u64.pow(attempt as u32);
                            sleep(Duration::from_millis(delay)).await;
                            continue;
                        }
                    }

                    // All retries exhausted or non-retryable error
                    let duration_ms = start.elapsed().as_millis() as u64;
                    return FetchResult {
                        source,
                        items: Vec::new(),
                        new_items: 0,
                        duration_ms,
                        error: Some(format!("{} after {} retries", e, retry_max)),
                    };
                }
            }
        }

        unreachable!()
    }

    async fn try_fetch(client: &Client, feed: &Feed) -> Result<Vec<FeedItem>, FetchError> {
        let mut request = client.get(&feed.url);

        for (key, value) in &feed.headers {
            request = request.header(key, value);
        }

        let response = request.send().await.map_err(|e| {
            if e.is_timeout() {
                FetchError::Timeout
            } else if e.is_connect() {
                FetchError::Connect
            } else {
                FetchError::Network(e.to_string())
            }
        })?;

        let status = response.status();
        if !status.is_success() {
            return Err(FetchError::Http(status));
        }

        let body = response.text().await.map_err(|e| FetchError::Network(e.to_string()))?;

        // Parse feed
        Parser::parse(&feed.name, &feed.feed_type, &body)
            .map_err(|e| FetchError::Parse(e))
    }

}

pub fn print_result(result: &FetchResult) {
    if let Some(error) = &result.error {
        eprintln!("  ✗ {:<25} — error: {}", result.source, error);
    } else {
        println!(
            "  ✓ {:<25} — {} items ({} new) in {}ms",
            result.source,
            result.items.len(),
            result.new_items,
            result.duration_ms
        );
    }
}

#[derive(Debug)]
enum FetchError {
    Timeout,
    Connect,
    Network(String),
    Http(reqwest::StatusCode),
    Parse(String),
}

impl std::fmt::Display for FetchError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            FetchError::Timeout => write!(f, "HTTP timeout"),
            FetchError::Connect => write!(f, "DNS resolution failure"),
            FetchError::Network(msg) => write!(f, "network error: {}", msg),
            FetchError::Http(status) => write!(f, "HTTP {}", status),
            FetchError::Parse(msg) => write!(f, "parse error: {}", msg),
        }
    }
}
